### PR TITLE
Added a bunch of utils, torch scaler, torch fbanks features, lightning training for SED etc. 

### PR DESCRIPTION
**Pull request by popcornell**
Still WIP, probably in conflict with https://github.com/DCASE-REPO/DESED_task/pull/5 . 
We have to decide what to merge first and then "harmonize" everything. 

1. Added torch datasets here for synth, val etc https://github.com/DCASE-REPO/DESED_task/blob/sam-lightning/desed/dataio/datasets.py
2. Better sampler class (my implementation from speechbrain):  https://github.com/DCASE-REPO/DESED_task/blob/sam-lightning/desed/dataio/sampler.py
3. on the fly gaussian noise and mixup augmentation with gpu https://github.com/DCASE-REPO/DESED_task/blob/sam-lightning/desed/gpu_augmentations.py
4. Torch scaler with more flexible arguments, torch scheduler, and modified ManyHotEncoder to take in account directly in that class for pooling performed in the DNN model. 
5. added requirements + linters. 
6.  TODO adding training loop once tested fully tonight.

### CODE REVIEW

File: recipes/dcase2021_task4_separation_SED/local/sed_training.py [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+
+        save_dir = os.path.join(self.logger.log_dir, "metrics_weak_val")
+        os.makedirs(save_dir, exist_ok=True)
+
+        _, _, weak_student_seg_macro, weak_student_seg_micro = log_sedeval_metrics(
+            self.val_buffer_student_weak, ground_truth, save_dir, self.current_epoch,
+        )
+        _, _, weak_teacher_seg_macro, weak_teacher_seg_micro = log_sedeval_metrics(
+            self.val_buffer_teacher_weak, ground_truth, save_dir, self.current_epoch,
+        )
```
Comment (turpaultn) [inline recipes/dcase2021_task4_separation_SED/local/sed_training.py:363]: The "weak score" is a usual macro f1-score using weak predictions and weak groundtruth.
I don't think this "trick" does the same 🤔 
See : https://github.com/turpaultn/dcase20_task4/blob/dd1d8bb434b08aab6a16a17b8445d4a80a750258/baseline/evaluation_measures.py#L372 
Sorry the code is a bit dirty (it returns the "score per class" so it needs to be averaged)

File: recipes/dcase2021_task4_separation_SED/local/utils.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import scipy
+import pandas as pd
+from pathlib import Path
+import os
+from desed.utils.evaluation_measures import compute_sed_eval_metrics
+
+
+def batched_decode_preds(
+    strong_preds, filenames, encoder, threshold=0.5, median_filter=7
```
Comment (turpaultn) [inline recipes/dcase2021_task4_separation_SED/local/utils.py:9]: median_filter is not set in the functions, so we always 7 (it matches the 0.45 s as long as the features are computed like this, so ok, but maybe we have to be careful for later expes)

File: recipes/dcase2021_task4_separation_SED/train_sed.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    sed_teacher = deepcopy(sed_student)
+
+    opt = torch.optim.Adam(sed_student.parameters(), 1e-8, betas=(0.9, 0.999))
+    exp_steps = config["training"]["n_epochs_warmup"] * epoch_len
+    exp_scheduler = {
+        "scheduler": ExponentialWarmup(opt, config["opt"]["lr"], exp_steps),
+        "interval": "step",
+    }
+
+    scaler = TorchScaler("instance", "minmax", (1, 2))
```
Comment (turpaultn) [inline recipes/dcase2021_task4_separation_SED/train_sed.py:119]: General thinking: 
For first experiment, do we need to put the scaler the same way as before (using the all dataset ? ) since I already make the experiment of how much it changes the performances in last year's code 🤔

**Review by turpaultn — CHANGES_REQUESTED** [review]
The only "requested_change" is really about changing desed to desed_task.
The other comments I guess Francesca can work on them to have a the "last year's version" of the baseline (mostly config to change, and the weak metric).

File: recipes/dcase2021_task4_separation_SED/confs/sed.yaml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  early_stop_patience: 10
+  accumulate_batches: 1
+  gradient_clip: 5
+  save_top_k: 1
+  median_window: 7
+  ema_factor: 0.999
+  self_sup_loss: mse
+  backend:
+  noise_snr: [12, 30]
+  validation_interval: 10
```
[the remark this answers is not in the document]
Comment (popcornell) [reply recipes/dcase2021_task4_separation_SED/confs/sed.yaml:18]: >  I guess the early stopping of 10 goes with that, so it is 10*10 the real early stopping number of epochs ? 

 Yes true

File: recipes/dcase2021_task4_separation_SED/local/utils.py (hunk 2 above)
Comment (popcornell) [reply recipes/dcase2021_task4_separation_SED/local/utils.py:9]: I ll add this in the config yaml

File: recipes/dcase2021_task4_separation_SED/local/sed_training.py (hunk 1 above)
Comment (popcornell) [reply recipes/dcase2021_task4_separation_SED/local/sed_training.py]: I changed it to standatd F1 macro averaged score over weak labels

File: recipes/dcase2021_task4_separation_SED/train_sed.py (hunk 3 above)
Comment (popcornell) [reply recipes/dcase2021_task4_separation_SED/train_sed.py:121]: Right now performance is decent even with min-max. For separation also it is useful to have this one. 
Maybe if we want to compare directly we need however the dataset scaler. Is the priority to make a comparison (which we can do later maybe) or deliver a baseline with last year plateau performance ?

**Comment by popcornell** [comment]
Merged changes from @RonFrancesca, addressed some issues in the @turpaultn  pointed out. Removed some warnings. 
I have yet to add weak labeled split in train_sed.py. I will copy the code you used which split the DataFrame.

File: recipes/dcase2021_task4_separation_SED/local/sed_training.py [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+
+        return Fbanks.take_log(mels)
+
+    def training_step(self, batch, batch_indx):
+
+        mixture, labels, padded_indxs = batch
+        indx_synth, indx_weak, indx_unlabelled = self.hparams["training"]["batch_size"]
+        mixture = self.feats(mixture)
+
+        mixture = add_noise(mixture, self.hparams["training"]["noise_snr"])
```
[the remark this answers is not in the document]
Comment (popcornell) [reply recipes/dcase2021_task4_separation_SED/local/sed_training.py:97]: Yes it it applied for all data. The SNR is so low that it does not impact I think. 
Actually I think it is better to add it cos right now we have zero padded sequences (masked loss is not used to discard the padded frames).

File: recipes/dcase2021_task4_separation_SED/local/sed_training.py [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+            "val/weak/student/loss_weak": loss_weak_student_on_weak,
+            "val/weak/teacher/loss_weak": loss_weak_teacher_on_weak,
+            "val/synth/student/loss_strong": loss_strong_student_on_synth,
+            "val/synth/teacher/loss_strong": loss_strong_teacher_on_synth,
+            "val/eval/student/loss_strong": loss_strong_student_on_eval,
+            "val/eval/teacher/loss_strong": loss_strong_teacher_on_eval,
+            "val/weak/student/segment_macro_F1": weak_student_seg_macro,
+            "val/weak/teacher/segment_macro_F1": weak_teacher_seg_macro,
+            "val/weak/student/segment_micro_F1": weak_student_seg_micro,
+            "val/weak/teacher/segment_micro_F1": weak_teacher_seg_micro,
```
[the remark this answers is not in the document]
Comment (popcornell) [reply recipes/dcase2021_task4_separation_SED/local/sed_training.py]: I would leave PSDS computation only for the test phase on dev-test

File: recipes/dcase2021_task4_separation_SED/local/sed_training.py [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+from collections import OrderedDict
+import pandas as pd
+from .utils import batched_decode_preds, log_sedeval_metrics, convert_to_event_based
+from desed.data_augm import add_noise, mixup, frame_shift
+from desed.features import Fbanks
+from pathlib import Path
+from desed.utils.torch_utils import nanmean, nantensor
+
+
+class DESED(pl.LightningModule):
```
[the remark this answers is not in the document]
Comment (popcornell) [reply recipes/dcase2021_task4_separation_SED/local/sed_training.py]: changed

File: setup.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+setup(
+    name="desed",
+    version="0.0.1",
+    description="Sound Event Detection and Separation in Domestic Environments.",
+    author="DCASE2021 Task 4 Organizers",
+    author_email="cornellsamuele@gmail.com",
+    license="MIT",
+    packages=["desed"],
+    python_requires=">=3.6",
+)
```
[the remark this answers is not in the document]
Comment (popcornell) [reply setup.py]: I ve put my email but actually maybe romain one is better

File: recipes/dcase2021_task4_separation_SED/local/sed_training.py [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+import pytorch_lightning as pl
+from collections import OrderedDict
+import pandas as pd
+from .utils import batched_decode_preds, log_sedeval_metrics
+
+
+class DESED(pl.LightningModule):
+    def __init__(
+        self,
+        hparams,
```
[the remark this answers is not in the document]
Comment (popcornell) [reply recipes/dcase2021_task4_separation_SED/local/sed_training.py:19]: @turpaultn actually however there is an advantage in having the yaml. It can be saved in the experiment directory and then it is easy to load to restore exactly the same hyperparameters. It is not so easy to do so with python. You have to import that.

File: requirements.txt [hunk 10]
```diff
@@ -0,0 +1,4 @@
+asteroid>=0.4.1
+dcase_util>=0.2.16
```
[the remark this answers is not in the document]
Comment (popcornell) [reply requirements.txt]: I use also features from asteroid however....

File: recipes/dcase2021_task4_separation_SED/local/sed_training.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (turpaultn) [reply recipes/dcase2021_task4_separation_SED/local/sed_training.py:19]: Well, it's just a matter of organising the python class, we could add a "save" function that saves the parameters in the yaml format, and a load to load from the yaml file.
Actually, even for the yaml file we have now, it could be a good solution because the python class could be helpful to keep track of the "used" parameters --> We could imagine the dictionary being filled when we do a "get" from the Configuration class, so all the unused parameters are not saved unnecessarily, what do you think ?

File: recipes/dcase2021_task4_separation_SED/train_sed.py (hunk 3 above)
Comment (turpaultn) [reply recipes/dcase2021_task4_separation_SED/train_sed.py:131]: The priority is to give the baseline, but it is important we have a list of everything that is changed from last year.
This scaler change, is mostly "useful" as you say so it is easy to use SSep with it, it seems we have an idea of how much it changes, so let's do the experiment later, but let's keep in mind we'll probably want to do it for the paper explaining the task + baseline. :)

File: recipes/dcase2021_task4_separation_SED/local/sed_training.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (turpaultn) [reply recipes/dcase2021_task4_separation_SED/local/sed_training.py]: That's not a problem no ? I've always assumed for our problem, the padding we're doing is small, it'll just force the model to predict "no class" at the moment where the sequence is padded

**Comment by turpaultn** [comment]
Merged in #9
